README.md
Gattica
=======
Gattica is an easy to use Gem for getting data from the Google Analytics API.  

Features
--------
* Supports: metrics, dimensions, sorting, filters, goals, and segments.
* Handles accounts with over 1000 profiles
* Returns data as: hash, json, CSV

[How to export Google Analytics data using Ruby](
http://www.seerinteractive.com/blog/google-analytics-data-export-api-with-rubygattica/2011/02/22/) (Links to my blog post on [Seer Interactive](http://www.seerinteractive.com))

<hr />

Quick Start
===========
Here are bare basics to get you up and running.

Installation
------------
Add Gattica to your Gemfile

    gem 'gattica', :git => 'git://github.com/chrisle/gattica.git'

Don't forget to bundle install:

    $ bundle install

Login, get a list of accounts, pick an account, and get data:

    # Include the gem
    require 'gattica'
    
    # Login
    ga = Gattica.new({ 
        :email => 'email@gmail.com', 
        :password => 'password'
    })

    # Get a list of accounts
    accounts = ga.accounts

    # Choose the first account
    ga.profile_id = accounts.first.profile_id

    # Get the data
    data = ga.get({ 
        :start_date   => '2011-01-01',
        :end_date     => '2011-04-01',
        :dimensions   => ['month', 'year'],
        :metrics      => ['visits', 'bounces'],
    })

    # Show the data
    puts data.inspect

<hr />

General Usage
=============

### Create your Gattica object

    ga = Gattica.new({ :email => 'email@gmail.com', :password => 'password' })
    puts ga.token   # => returns a big alpha-numeric string
    
### Query for accounts you have access to

    # Retrieve a list of accounts
    accounts = ga.accounts

    # Show information about accounts
    puts "---------------------------------"
    puts "Available profiles: " + accounts.count.to_s
    accounts.each do |account|
      puts "   --> " + account.title
      puts "   last updated: " + account.updated.inspect
      puts "   web property: " + account.web_property_id
      puts "     profile id: " + account.profile_id.inspect
      puts "          goals: " + account.goals.count.inspect
    end

### Set which profile Gattica needs to use

    # Tell Gattica to query profile ID 5555555
    ga.profile_id = 5555555 

### Get data from Google Analytics
    
The Get method will get data from Google Analytics and return Gattica::DataSet type.  

* Dates must be in 'YYYY-MM-DD' format.  
* Dimensions and metrics can be gotten from [Google Analytics Dimensions & Metrics Reference](http://code.google.com/apis/analytics/docs/gdata/gdataReferenceDimensionsMetrics.html)
* You do not need to use "ga:" at the beginning of the strings.

Here's an example:

    # Get the number of visitors by month from Jan 1st to April 1st.
    data = ga.get({ 
        :start_date   => '2011-01-01',
        :end_date     => '2011-04-01',
        :dimensions   => ['month', 'year'],
        :metrics      => ['visitors']
    })

<hr />

Using Dimension & Metrics
=========================

Here are some additional examples that illustrate different things you can do with dimensions and metrics.


### Sorting

    # Sorting by number of visits in descending order (most visits at the top)
    data = ga.get({ 
        :start_date   => '2011-01-01',
        :end_date     => '2011-04-01',
        :dimensions   => ['month', 'year'],
        :metrics      => ['visits']
        :sort         => ['-visits'],
    })

### Limiting results

    # Limit the number of results to 25.
    data = ga.get({ 
        :start_date   => '2011-01-01',
        :end_date     => '2011-04-01',
        :dimensions   => ['month', 'year'],
        :metrics      => ['visits']
        :max_results  => 25 
    })

### Results as a Hash

    my_hash = data.to_h['points']

    # => 
    #   [{
    #     "xml"         => "<entry gd:etag=\"W/&quot;....  </entry>", 
    #     "id"          => "http://www.google.com/analytics/feeds/data?...", 
    #     "updated"     => Thu, 31 Mar 2011 17:00:00 -0700, 
    #     "title"       => "ga:month=01 | ga:year=2011", 
    #     "dimensions"  => [{:month=>"01"}, {:year=>"2011"}], 
    #     "metrics"     => [{:visitors=>6}]
    #   },
    #   {
    #     "xml"         => ...
    #     "id"          => ...
    #     "updated"     => ...
    #     ...
    #   }]


### JSON formatted string

    # Return data as a json string. (Useful for NoSQL databases)
    my_json = data.to_h['points'].to_json

    # => 
    #   "[{
    #       \"xml\":\"<entry> .... </entry>\",
    #       \"id\":\"http://www.google.com/analytics/feeds/data? ...",
    #       \"updated\":\"2011-03-31T17:00:00-07:00\",
    #       \"title\":\"ga:month=01 | ga:year=2011\",
    #       \"dimensions\":[{\"month\":\"01\"},{\"year\":\"2011\"}],
    #       \"metrics\":[{\"visitors\":6}]
    #     },
    #     { 
    #       \"xml\":\"<entry> .... </entry>\",
    #       \"id\":\"http://www.google.com/analytics/feeds/data? ...",
    #       ...
    #   }]"
    
### CSV formatted string

    # Return the data in CSV format.  (Useful for using in Excel.)

    # Short CSV will only return your dimensions and metrics:
    short_csv = data.to_csv(:short)   
    
    # => "month,year,visitors\n\n01,2011, ...."

    # Long  CSV will get you a few additional columns:
    long_csv = data.to_csv            
    
    # => "id,updated,title,month,year,visitors\n\nhttp:// ..."


### DIY formatting

    # You can work directly with the 'point' method to return data.
    data.points.each do |data_point|
      month = data_point.dimensions.detect { |dim| dim.key == :month }.value
      year = data_point.dimensions.detect { |dim| dim.key == :year }.value
      visitors = data_point.metrics.detect { |metric| metric.key == :visitors }.value
      puts "#{month}/#{year} got #{visitors} visitors"
    end

    # => 
    #   01/2011 got 34552 visitors
    #   02/2011 got 36732 visitors
    #   03/2011 got 45642 visitors
    #   04/2011 got 44456 visitors

<hr />

Using Filter, Goals, and Segments
=========================

Learn more about filters: [Google Data feed filtering reference](http://code.google.com/apis/analytics/docs/gdata/gdataReference.html#filtering)


### Get profiles with goals

    # Get all the profiles that have goals
    profiles_with_goals = accounts.select { |account| account.goals.count > 0 }

    # => 
    #   [{
    #     "id"                => "http://www.google.com/analytics/feeds/accounts/ga:...",
    #     "updated"           => Mon, 16 May 2011 16:40:30 -0700, 
    #     "title"             => "Profile Title", 
    #     "table_id"          => "ga:123456", 
    #     "account_id"        => 123456, 
    #     "account_name"      => "Account name", 
    #     "profile_id"        =>  123456, 
    #     "web_property_id"   => "UA-123456-3", 
    #     "goals"=>[{
    #         :active   => "true", 
    #         :name     => "Goal name", 
    #         :number   => 1, 
    #         :value    => 0.0
    #     }]
    #   }, 
    #   {
    #     "id"                => "http://www.google.com/analytics/feeds/accounts/ga:...",
    #     "updated"           => Mon, 16 May 2011 16:40:30 -0700, 
    #     "title"             => "Profile Title", 
    #     ...
    #   }]

### List available segments

    # Get all the segments that are available to you
    segments = ga.segments

    # Segments with negative gaid are default segments from Google. Segments
    # with positive gaid numbers are custom segments that you created.
    # =>
    #   [{
    #     "id"          => "gaid::-1", 
    #     "name"        => "All Visits", 
    #     "definition"  => " "
    #   }, 
    #   {
    #     "id"          => "gaid::-2", 
    #     "name"        => "New Visitors", 
    #     "definition"  => "ga:visitorType==New Visitor"
    #   }, 
    #   {
    #     "id"          => ... # more default segments
    #     "name"        => ...
    #     "definition"  => ...
    #   },
    #   {
    #     "id"          => "gaid::12345678", 
    #     "name"        => "Name of segment", 
    #     "definition"  => "ga:keyword=...."
    #   }, 
    #   {
    #     "id"          => ... # more custom segments
    #     "name"        => ...
    #     "definition"  => ...
    #   }]

### Query by segment

    # Return visits and bounces for mobile traffic 
    # (Google's default user segment gaid::-11)
    
    mobile_traffic = ga.get({ 
      :start_date   => '2011-01-01', 
      :end_date     => '2011-02-01', 
      :dimensions   => ['month', 'year'],
      :metrics      => ['visits', 'bounces'],
      :segment      => ['gaid::-11']
    })

### Filtering

Filters are boolean expressions in strings. Here's an example of an equality:

    # Filter by Firefox users
    firefox_users = ga.get({
      :start_date   => '2010-01-01', 
      :end_date     => '2011-01-01',
      :dimensions   => ['month', 'year'],
      :metrics      => ['visits', 'bounces'],
      :filters      => ['browser == Firefox']
    })
    
Here's an example of greater-than:
    
    # Filter where visits is >= 10000
    lots_of_visits = ga.get({
      :start_date   => '2010-01-01', 
      :end_date     => '2011-02-01',
      :dimensions   => ['month', 'year'],
      :metrics      => ['visits', 'bounces'],
      :filters      => ['visits >= 10000']
    })
    
Multiple filters is an array.  Currently, they are only joined by 'AND'.

    # Firefox users and visits >= 10000
    firefox_users_with_many_pageviews = ga.get({
      :start_date   => '2010-01-01', 
      :end_date     => '2011-02-01',
      :dimensions   => ['month', 'year'],
      :metrics      => ['visits', 'bounces'],
      :filters      => ['browser == Firefox', 'visits >= 10000']
    })


<hr />

Even More Examples!
===============

### Top 25 keywords that drove traffic

Output the top 25 keywords that drove traffic to your website in the first quarter of 2011.

    # Get the top 25 keywords that drove traffic
    data = ga.get({ 
      :start_date => '2011-01-01',
      :end_date => '2011-04-01',
      :dimensions => ['keyword'],
      :metrics => ['visits'],
      :sort => ['-visits'],
      :max_results => 25 
    })
    
    # Output our results
    data.points.each do |data_point|
      kw = data_point.dimensions.detect { |dim| dim.key == :keyword }.value
      visits = data_point.metrics.detect { |metric| metric.key == :visits }.value
      puts "#{visits} visits => '#{kw}'"
    end

    # =>
    #   19667 visits => '(not set)'
    #   1677 visits => 'keyword 1'
    #   178 visits => 'keyword 2'
    #   165 visits => 'keyword 3'
    #   161 visits => 'keyword 4'
    #   112 visits => 'keyword 5'
    #   105 visits => 'seo company reviews'
    #   ...
    
<hr />

Additional Options & Settings
=============================

Setting HTTP timeout
--------------------

If you have a lot of profiles in your account (like 1000+ profiles) querying for accounts may take over a minute.  Net::HTTP will timeout and an exception will be raised.

To avoid this, specify a timeout when you instantiate the Gattica object:

    ga = Gattica.new({ 
        :email => 'email@gmail.com', 
        :password => 'password',
        :timeout => 600  # Set timeout for 10 minutes!
    })

The default timeout is 300 seconds (5 minutes). Change the default in: lib/gattica/settings.rb

For reference 1000 profiles with 2-5 goals each takes around 90-120 seconds.

Reusing a session token
-----------------------

You can reuse an older session if you still have the token string.  Google recommends doing this to avoid authenticating over and over.

  
    my_token = ga.token # => 'DSasdf94...'
    
    # Sometime later, you can initialize Gattica with the same token
    ga = Gattica.new({ :token => my_token })

If your token times out, you will need to re-authenticate.

Specifying your own headers
---------------------------

Google expects a special header in all HTTP requests called 'Authorization'.  Gattica handles this header automatically.  If you want to specify your own you can do that when you instantiate Gattica:

    ga = Gattica.new({
        :token => 'DSasdf94...', 
        :headers => {'My-Special-Header':'my_custom_value'}
    })
        
<hr />

History
=======

Version history
---------------
### 0.4.4
  * Added a configuration file to unit tests
  * Removed version.rb.  Not needed. (thanks John McGrath see: github.com/john)
  * Migrated examples and rewrote README file

### 0.4.3
  * FIXED: Typo in start-index parameter
  * Refactored Engine class into it's own file.
  * Began to re-style code to wrap at 80 characters
  * Added some unit tests

### 0.4.2
  * Added Ruby 1.9 support (Thanks @mathieuravaux https://github.com/mathieuravaux)
  * Uses hpricot 0.8.4 now.  0.8.3 segfaults.
  * Added ability to change the timeout when requesting analytics from Google
  * Added the ability to use max_results

### 0.3.2.scottp
  * scottp Added Analytics API v2 header, and basic support for "segment" argument.

### 0.3.2
  * er1c updated to use standard Ruby CSV library

### 0.3.0
  * Support for filters (filters are all AND'ed together, no OR yet)

### 0.2.1 
  * More robust error checking on HTTP calls
  * Added to_xml to get raw XML output from Google
  
### 0.2.0 / 2009-04-27
  * Changed initialization format: pass a hash of options rather than individual email, password and profile_id
  * Can initialize with a valid token and use that instead of requiring email/password each time
  * Can initialize with your own logger object instead of having to use the default (useful if you're using with Rails, initialize with RAILS_DEFAULT_LOGGER)
  * Show error if token is invalid or expired (Google returns a 401 on any HTTP call)
  * Started tests

### 0.1.4 / 2009-04-22
  * Another attempt at getting the gem to build on github

### 0.1.3 / 2009-04-22
  * Getting gem to build on github

### 0.1.2 / 2009-04-22
  * Updated readme and examples, better documentation throughout

### 0.1.1 / 2009-04-22
  * When outputting as CSV, surround each piece of data with double quotes (appears pretty common for various properties (like Browser name) to contain commas

### 0.1.0 / 2009-03-26
  * Basic functionality working good. Can't use filters yet.
  

Maintainer history
------------------
  * [Rob Cameron](https://github.com/activenetwork/gattica) (2010)
  * [Mike Rumble](https://github.com/rumble/gattica) (2010)
  * [Chris Le](https://github.com/chrisle/gattica) (Current)
  

History.txt
== 0.4.3
  * FIXED: Typo in start-index parameter
  * Refactored Engine class into it's own file.
  * Began to re-style code to wrap at 80 characters
  * Added some unit tests

== 0.4.2
  * Added Ruby 1.9 support (Thanks @mathieuravaux https://github.com/mathieuravaux)
  * Uses hpricot 0.8.4 now.  0.8.3 segfaults.
  * Added ability to change the timeout when requesting analytics from Google
  * Added the ability to use max_results

== 0.3.2.scottp
  * scottp Added Analytics API v2 header, and basic support for "segment" argument.

== 0.3.2
  * er1c updated to use standard Ruby CSV library

== 0.3.0
  * Support for filters (filters are all AND'ed together, no OR yet)

== 0.2.1 
  * More robust error checking on HTTP calls
  * Added to_xml to get raw XML output from Google
  
== 0.2.0 / 2009-04-27
  * Changed initialization format: pass a hash of options rather than individual email, password and profile_id
  * Can initialize with a valid token and use that instead of requiring email/password each time
  * Can initialize with your own logger object instead of having to use the default (useful if you're using with Rails, initialize with RAILS_DEFAULT_LOGGER)
  * Show error if token is invalid or expired (Google returns a 401 on any HTTP call)
  * Started tests

== 0.1.4 / 2009-04-22
  * Another attempt at getting the gem to build on github

== 0.1.3 / 2009-04-22
  * Getting gem to build on github

== 0.1.2 / 2009-04-22
  * Updated readme and examples, better documentation throughout

== 0.1.1 / 2009-04-22
  * When outputting as CSV, surround each piece of data with double quotes (appears pretty common for various properties (like Browser name) to contain commas

== 0.1.0 / 2009-03-26
  * Basic functionality working good. Can't use filters yet.

test/test_auth.rb
require File.expand_path('../helper', __FILE__)

class TestAuth < Test::Unit::TestCase
  def setup
    
  end
  
  def test_truth
    assert true
  end
  
end

test/test_user.rb
require_relative 'helper'
 
class TestUser < Test::Unit::TestCase
  def setup
    
  end
  
  def test_can_create_user
    assert Gattica::User.new('anonymous@anon.com','none')
  end
  
  def test_invalid_email
    assert_raise GatticaError::InvalidEmail do Gattica::User.new('','') end
    assert_raise ArgumentError do Gattica::User.new('') end
    assert_raise GatticaError::InvalidEmail do Gattica::User.new('anonymous','none') end
    assert_raise GatticaError::InvalidEmail do Gattica::User.new('anonymous@asdfcom','none') end
  end
    
  def test_invalid_password
    assert_raise GatticaError::InvalidPassword do Gattica::User.new('anonymous@anon.com','') end
    assert_raise ArgumentError do Gattica::User.new('anonymous@anon.com') end
  end
  
end


lib/gattica.rb
$:.unshift File.dirname(__FILE__) # for use/testing when no gem is installed

require 'net/http'
require 'net/https'
require 'uri'
require 'cgi'
require 'logger'
require 'rubygems'
require 'hpricot'
require 'yaml'

require 'gattica/engine'
require 'gattica/settings'
require 'gattica/version'
require 'gattica/core_extensions'
require 'gattica/convertible'
require 'gattica/exceptions'
require 'gattica/user'
require 'gattica/auth'
require 'gattica/account'
require 'gattica/data_set'
require 'gattica/data_point'
require 'gattica/segment'

# Gattica is a Ruby library for talking to the Google Analytics API.
# Please see the README for usage docs.
module Gattica
  
  # Creates a new instance of Gattica::Engine and gets us going. Please see the README for usage docs.
  #   ga = Gattica.new({:email => 'anonymous@anon.com', :password => 'password, :profile_id => 123456 })
  
  def self.new(*args)
    Engine.new(*args)
  end
  
end


lib/gattica/account.rb
require 'rubygems'
require 'hpricot'

module Gattica
  
  # Represents an account that an authenticated user has access to
  
  class Account
    
    include Convertible
    
    attr_reader :id, :updated, :title, :table_id, :account_id, :account_name,
                :profile_id, :web_property_id, :goals
  
    def initialize(xml)
      @id = xml.at(:id).inner_html
      @updated = DateTime.parse(xml.at(:updated).inner_html)
      @title = xml.at(:title).inner_html
      @table_id = xml.at('dxp:tableid').inner_html
      @account_id = xml.at("dxp:property[@name='ga:accountId']").attributes['value'].to_i
      @account_name = xml.at("dxp:property[@name='ga:accountName']").attributes['value']
      @profile_id = xml.at("dxp:property[@name='ga:profileId']").attributes['value'].to_i
      @web_property_id = xml.at("dxp:property[@name='ga:webPropertyId']").attributes['value']
      @goals = xml.search('ga:goal').collect do |goal| {
        :active => goal.attributes['active'],
        :name => goal.attributes['name'],
        :number => goal.attributes['number'].to_i,
        :value => goal.attributes['value'].to_f,
      }
      end
    end
    
  end
end


lib/gattica/settings.rb
module Gattica
  module Settings

    # Google server to connect to
    #
    # You may need to change this if you use a proxy server.  However, I have not
    # tested using a proxy.... YMMV
    SERVER = 'www.google.com'

    # Use a secure connection?
    #
    # If set to true, Gattica will use SSL_PORT or else it will use NON_SSL_PORT.
    #
    # === Example:
    # +USE_SSL = true+::  # Always use a secure connection
    USE_SSL = true
    SSL_PORT = 443
    NON_SSL_PORT = 80

    # Net::HTTP timeout in seconds
    #
    # Increase the timeout setting if you have over 100 profiles.
    #
    # === Example:
    # +TIMEOUT = 300+::   # Increase timeout to 5 minutes.
    TIMEOUT = 300

    DEFAULT_ARGS = {
        :start_date => nil,
        :end_date => nil,
        :dimensions => [],
        :metrics => [],
        :filters => [],
        :sort => []
    }

    DEFAULT_OPTIONS = {
        :email => nil,        # eg: 'email@gmail.com'
        :password => nil,     # eg: '$up3r_$ekret'
        :token => nil,
        :profile_id => nil,
        :debug => false,
        :headers => {},
        :logger => Logger.new(STDOUT)
    }

    FILTER_METRIC_OPERATORS = %w{ == != > < >= <= }
    FILTER_DIMENSION_OPERATORS = %w{ == != =~ !~ =@ ~@ }
    
  end
end

lib/gattica/engine.rb
module Gattica
  class Engine

    attr_reader :user
    attr_accessor :profile_id, :token

    # Initialize Gattica using username/password or token.
    #
    # = Options:
    # To change the defaults see link:settings.rb
    # +:debug+::        Send debug info to the logger (default is false)
    # +:email+::        Your email/login for Google Analytics
    # +:headers+::      Add additional HTTP headers (default is {} )
    # +:logger+::       Logger to use (default is STDOUT)
    # +:password+::     Your password for Google Analytics
    # +:profile_id+::   Use this Google Analytics profile_id (default is nil)
    # +:timeout+::      Set Net:HTTP timeout in seconds (default is 300)
    # +:token+::        Use an authentication token you received before
    def initialize(options={})
      @options = Settings::DEFAULT_OPTIONS.merge(options)
      handle_init_options(@options)
      create_http_connection()
      check_init_auth_requirements()
      # TODO: check that the user has access to the specified profile and show an error here rather than wait for Google to respond with a message
    end

    # Returns the list of accounts the user has access to. A user may have multiple accounts on Google Analytics
    # and each account may have multiple profiles. You need the profile_id in order to get info from GA. If you
    # don't know the profile_id then use this method to get a list of all them. Then set the profile_id of your
    # instance and you can make regular calls from then on.
    #
    #   ga = Gattica.new({:email => 'johndoe@google.com', :password => 'password'})
    #   ga.accounts
    #   # you parse through the accounts to find the profile_id you need
    #   ga.profile_id = 12345678
    #   # now you can perform a regular search, see Gattica::Engine#get
    #
    # If you pass in a profile id when you instantiate Gattica::Search then you won't need to
    # get the accounts and find a profile_id - you apparently already know it!
    #
    # See Gattica::Engine#get to see how to get some data.

    def accounts
      # if we haven't retrieved the user's accounts yet, get them now and save them
      if @user_accounts.nil?
        data = request_default_account_feed
        xml = Hpricot(data)
        @user_accounts = xml.search(:entry).collect { |entry| Account.new(entry) }
      end
      return @user_accounts
    end

    # Returns the list of segments available to the authenticated user.
    #
    # == Usage
    #   ga = Gattica.new({:email => 'johndoe@google.com', :password => 'password'})
    #   ga.segments                       # Look up segment id
    #   my_gaid = 'gaid::-5'              # Non-paid Search Traffic
    #   ga.profile_id = 12345678          # Set our profile ID
    #
    #   gs.get({ :start_date => '2008-01-01',
    #            :end_date => '2008-02-01',
    #            :dimensions => 'month',
    #            :metrics => 'views',
    #            :segment => my_gaid })

    def segments
      if @user_segments.nil?
        data = request_default_account_feed
        xml = Hpricot(data)
        @user_segments = xml.search("dxp:segment").collect { |s| Segment.new(s) }
      end
      return @user_segments
    end

    # This is the method that performs the actual request to get data.
    #
    # == Usage
    #
    #   gs = Gattica.new({:email => 'johndoe@google.com', :password => 'password', :profile_id => 123456})
    #   gs.get({ :start_date => '2008-01-01',
    #            :end_date => '2008-02-01',
    #            :dimensions => 'browser',
    #            :metrics => 'pageviews',
    #            :sort => 'pageviews',
    #            :filters => ['browser == Firefox']})
    #
    # == Input
    #
    # When calling +get+ you'll pass in a hash of options. For a description of what these mean to
    # Google Analytics, see http://code.google.com/apis/analytics/docs
    #
    # Required values are:
    #
    # * +start_date+ => Beginning of the date range to search within
    # * +end_date+ => End of the date range to search within
    #
    # Optional values are:
    #
    # * +dimensions+ => an array of GA dimensions (without the ga: prefix)
    # * +metrics+ => an array of GA metrics (without the ga: prefix)
    # * +filter+ => an array of GA dimensions/metrics you want to filter by (without the ga: prefix)
    # * +sort+ => an array of GA dimensions/metrics you want to sort by (without the ga: prefix)
    #
    # == Exceptions
    #
    # If a user doesn't have access to the +profile_id+ you specified, you'll receive an error.
    # Likewise, if you attempt to access a dimension or metric that doesn't exist, you'll get an
    # error back from Google Analytics telling you so.

    def get(args={})
      args = validate_and_clean(Settings::DEFAULT_ARGS.merge(args))
      query_string = build_query_string(args,@profile_id)
      @logger.debug(query_string) if @debug
      data = do_http_get("/analytics/feeds/data?#{query_string}")
      #data = do_http_get("/analytics/feeds/data?ids=ga%3A915568&metrics=ga%3Avisits&segment=gaid%3A%3A-7&start-date=2010-03-29&end-date=2010-03-29&max-results=50")
      return DataSet.new(Hpricot.XML(data))
    end


    # Since google wants the token to appear in any HTTP call's header, we have to set that header
    # again any time @token is changed so we override the default writer (note that you need to set
    # @token with self.token= instead of @token=)

    def token=(token)
      @token = token
      set_http_headers
    end

    ######################################################################
    private

    # Gets the default account feed from Google
    def request_default_account_feed
      if @default_account_feed.nil?
        @default_account_feed = do_http_get('/analytics/feeds/accounts/default')
      end
      return @default_account_feed
    end

    # Does the work of making HTTP calls and then going through a suite of tests on the response to make
    # sure it's valid and not an error

    def do_http_get(query_string)
      response, data = @http.get(query_string, @headers)

      # error checking
      if response.code != '200'
        case response.code
        when '400'
          raise GatticaError::AnalyticsError, response.body + " (status code: #{response.code})"
        when '401'
          raise GatticaError::InvalidToken, "Your authorization token is invalid or has expired (status code: #{response.code})"
        else  # some other unknown error
          raise GatticaError::UnknownAnalyticsError, response.body + " (status code: #{response.code})"
        end
      end

      return data
    end


    # Sets up the HTTP headers that Google expects (this is called any time @token is set either by Gattica
    # or manually by the user since the header must include the token)
    def set_http_headers
      @headers['Authorization'] = "GoogleLogin auth=#{@token}"
      @headers['GData-Version']= '2'
    end


    # Creates a valid query string for GA
    def build_query_string(args,profile)
      output = "ids=ga:#{profile}&start-date=#{args[:start_date]}&end-date=#{args[:end_date]}"
      if (start_index = args[:start_index].to_i) > 0
        output += "&start-index=#{start_index}"
      end
      unless args[:dimensions].empty?
        output += '&dimensions=' + args[:dimensions].collect do |dimension|
          "ga:#{dimension}"
        end.join(',')
      end
      unless args[:metrics].empty?
        output += '&metrics=' + args[:metrics].collect do |metric|
          "ga:#{metric}"
        end.join(',')
      end
      unless args[:sort].empty?
        output += '&sort=' + args[:sort].collect do |sort|
          sort[0..0] == '-' ? "-ga:#{sort[1..-1]}" : "ga:#{sort}"  # if the first character is a dash, move it before the ga:
        end.join(',')
      end
      unless args[:segment].nil?
        output += "&segment=#{args[:segment]}"
      end
      unless args[:max_results].nil?
        output += "&max-results=#{args[:max_results]}"
      end

      # TODO: update so that in regular expression filters (=~ and !~), any initial special characters in the regular expression aren't also picked up as part of the operator (doesn't cause a problem, but just feels dirty)
      unless args[:filters].empty?    # filters are a little more complicated because they can have all kinds of modifiers
        output += '&filters=' + args[:filters].collect do |filter|
          match, name, operator, expression = *filter.match(/^(\w*)\s*([=!<>~@]*)\s*(.*)$/)           # splat the resulting Match object to pull out the parts automatically
          unless name.empty? || operator.empty? || expression.empty?                      # make sure they all contain something
            "ga:#{name}#{CGI::escape(operator.gsub(/ /,''))}#{CGI::escape(expression)}"   # remove any whitespace from the operator before output
          else
            raise GatticaError::InvalidFilter, "The filter '#{filter}' is invalid. Filters should look like 'browser == Firefox' or 'browser==Firefox'"
          end
        end.join(';')
      end
      return output
    end


    # Validates that the args passed to +get+ are valid
    def validate_and_clean(args)

      raise GatticaError::MissingStartDate, ':start_date is required' if args[:start_date].nil? || args[:start_date].empty?
      raise GatticaError::MissingEndDate, ':end_date is required' if args[:end_date].nil? || args[:end_date].empty?
      raise GatticaError::TooManyDimensions, 'You can only have a maximum of 7 dimensions' if args[:dimensions] && (args[:dimensions].is_a?(Array) && args[:dimensions].length > 7)
      raise GatticaError::TooManyMetrics, 'You can only have a maximum of 10 metrics' if args[:metrics] && (args[:metrics].is_a?(Array) && args[:metrics].length > 10)

      possible = args[:dimensions] + args[:metrics]

      # make sure that the user is only trying to sort fields that they've previously included with dimensions and metrics
      if args[:sort]
        missing = args[:sort].find_all do |arg|
          !possible.include? arg.gsub(/^-/,'')    # remove possible minuses from any sort params
        end
        unless missing.empty?
          raise GatticaError::InvalidSort, "You are trying to sort by fields that are not in the available dimensions or metrics: #{missing.join(', ')}"
        end
      end

      # make sure that the user is only trying to filter fields that are in dimensions or metrics
      if args[:filters]
        missing = args[:filters].find_all do |arg|
          !possible.include? arg.match(/^\w*/).to_s    # get the name of the filter and compare
        end
        unless missing.empty?
          raise GatticaError::InvalidSort, "You are trying to filter by fields that are not in the available dimensions or metrics: #{missing.join(', ')}"
        end
      end

      return args
    end

    def create_http_connection
      port = Settings::USE_SSL ? Settings::SSL_PORT : Settings::NON_SSL_PORT
      @http = Net::HTTP.new(Settings::SERVER, port)
      @http.use_ssl = Settings::USE_SSL
      @http.set_debug_output $stdout if @options[:debug]
      defined? @options[:timeout] ?
          (@http.read_timeout = @options[:timeout]) :
          (@http.read_timeout = Settings::TIMEOUT)
    end

    # Sets instance variables from options given during initialization and
    def handle_init_options(options)
      @logger = options[:logger]
      @profile_id = options[:profile_id]
      @user_accounts = nil # filled in later if the user ever calls Gattica::Engine#accounts
      @user_segments = nil
      @headers = { }.merge(options[:headers]) # headers used for any HTTP requests (Google requires a special 'Authorization' header which is set any time @token is set)
      @default_account_feed = nil
    end

    def check_init_auth_requirements
      if @options[:email] && @options[:password]
        @user = User.new(@options[:email], @options[:password])
        @auth = Auth.new(@http, user)
        self.token = @auth.tokens[:auth]
      elsif @options[:token]
        self.token = @options[:token]
      else
        raise GatticaError::NoLoginOrToken, 'An email and password or an authentication token is required to initialize Gattica.'
      end
    end

  end
end
